runner.go

```diff
@@ -124,49 +125,65 @@ func (r *runner) getWeightSum() (weightSum int) {
 func (r *runner) spawnWorkers(spawnCount int, quit chan bool, hatchCompleteFunc func()) {
 	log.Println("Hatching and swarming", spawnCount, "clients at the rate", r.hatchRate, "clients/s...")
 
-	weightSum := r.getWeightSum()
-	for _, task := range r.tasks {
-		percent := float64(task.Weight) / float64(weightSum)
-		amount := int(round(float64(spawnCount)*percent, .5, 0))
-
-		if weightSum == 0 {
-			amount = int(float64(spawnCount) / float64(len(r.tasks)))
+	defer func() {
+		if hatchCompleteFunc != nil {
+			hatchCompleteFunc()
 		}
+	}()
 
-		for i := 1; i <= amount; i++ {
-			sleepTime := time.Duration(1000000/r.hatchRate) * time.Microsecond
-			time.Sleep(sleepTime)
-
-			select {
-			case <-quit:
-				// quit hatching goroutine
-				return
-			default:
-				atomic.AddInt32(&r.numClients, 1)
-				go func(fn func()) {
-					for {
-						select {
-						case <-quit:
-							return
-						default:
-							if r.rateLimitEnabled {
-								blocked := r.rateLimiter.Acquire()
-								if !blocked {
-									r.safeRun(fn)
-								}
-							} else {
-								r.safeRun(fn)
+	for i := 1; i <= spawnCount; i++ {
+		sleepTime := time.Duration(1000000/r.hatchRate) * time.Microsecond
+		time.Sleep(sleepTime)
+
+		select {
+		case <-quit:
+			// quit hatching goroutine
+			return
+		default:
+			atomic.AddInt32(&r.numClients, 1)
+			go func() {
+				for {
+					task := r.getRandomTask()
+					select {
+					case <-quit:
+						return
+					default:
+						if r.rateLimitEnabled {
+							blocked := r.rateLimiter.Acquire()
+							if !blocked {
+								r.safeRun(task.Fn)
 							}
+						} else {
+							r.safeRun(task.Fn)
 						}
 					}
-				}(task.Fn)
-			}
+				}
+			}()
 		}
 	}
+}
+
+func (r *runner) getRandomTask() *Task {
+	totalWeight := r.getWeightSum()
+	if totalWeight == 0 {
+		if len(r.tasks) > 0 {
+			return r.tasks[0]
+		}
+		return nil
+	}
 
-	if hatchCompleteFunc != nil {
-		hatchCompleteFunc()
+	randSrc := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
This doesn't have much affect on my usecase (because we manually add 100ms to 10s waits between tasks), but maybe it doesn't make sense to need to create a new random object on each `getRandomTask` call, but cache that off somewhere?